lxd/patches.go

```diff
@@ -952,40 +1007,249 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+				return err
+			}
+
+			// Check for a root disk device entry
+			k, _ := containerGetRootDiskDevice(p.Devices)
+			if k != "" {
+				// If found, set the pool property
+				p.Devices[k]["pool"] = poolName
+			} else if k == "" && pName == "default" {
+				// The default profile should have a valid root
+				// disk device entry.
+				rootDev := map[string]string{}
+				rootDev["type"] = "disk"
+				rootDev["path"] = "/"
+				rootDev["pool"] = poolName
+				if p.Devices == nil {
+					p.Devices = map[string]map[string]string{}
+				}
+
+				// Make sure that we do not overwrite a device the user
+				// is currently using under the name "root".
+				rootDevName := "root"
+				for i := 0; i < 100; i++ {
+					if p.Devices[rootDevName] == nil {
+						break
+					}
+					rootDevName = fmt.Sprintf("root%d", i)
+					continue
+				}
+				p.Devices["root"] = rootDev
+			}
+
+			// This is nasty, but we need to clear the profiles config and
+			// devices in order to add the new root device including the
+			// newly added storage pool.
+			tx, err := dbBegin(d.db)
+			if err != nil {
+				return err
+			}
+
+			err = dbProfileConfigClear(tx, pID)
+			if err != nil {
+				shared.LogErrorf("Failed to clear old profile configuration: %s.", err)
+				tx.Rollback()
+				continue
+			}
+
+			err = dbProfileConfigAdd(tx, pID, p.Config)
+			if err != nil {
+				shared.LogErrorf("Failed to add new profile configuration: %s.", err)
+				tx.Rollback()
+				continue
+			}
+
+			err = dbDevicesAdd(tx, "profile", pID, p.Devices)
+			if err != nil {
+				shared.LogErrorf("Failed to add new profile profile root disk device: %s.", err)
+				tx.Rollback()
+				continue
+			}
+
+			err = tx.Commit()
+			if err != nil {
+				shared.LogErrorf("Failed to commit database transaction: %s.", err)
+				tx.Rollback()
+				continue
+			}
+		}
+	}
+
+	// When no default profile is detected or some containers do not rely on
+	// the default profile for their root disk device, these containers will
+	// be given a valid local root disk device."
+	for _, ct := range allcontainers {
+		c, err := containerLoadByName(d, ct)
+		if err != nil {
+			continue
+		}
+
+		args := containerArgs{
+			Architecture: c.Architecture(),
+			Config:       c.LocalConfig(),
+			Ephemeral:    c.IsEphemeral(),
+			CreationDate: c.CreationDate(),
+			LastUsedDate: c.LastUsedDate(),
+			Name:         c.Name(),
+			Profiles:     c.Profiles(),
 		}
+
+		if c.IsSnapshot() {
+			args.Ctype = cTypeSnapshot
+		} else {
+			args.Ctype = cTypeRegular
+		}
+
+		// Check expanded devices for a valid root entry. If it exists,
+		// we skip this container.
+		expandedDevices := c.ExpandedDevices()
+		k, _ := containerGetRootDiskDevice(expandedDevices)
+		if k != "" && expandedDevices[k]["pool"] != "" {
+			// On partial upgrade the container might already have a
+			// valid root disk device entry.
+			if expandedDevices[k]["pool"] == poolName {
+				continue
```

Comment:
Nevermind, I see you skip the upgrade if there are more than one pools defined. That's not perfect but I guess we can tell those users how to fix their profiles if they got that far.